### PR TITLE
✨ Remove deprecated go get from Makefile templates

### PR DESCRIPTION
**Pull request by ryantking**
Signed-off-by: Ryan King <ryking@redhat.com>

In go 1.18, using `go get` to install go binaries will be removed from
in favor of using `go install`. This commit updates the Makefile
templates to use alternative installation methods to install the tools
used at generation.

Since `go install` adds extra constraints on what go packages can be
installed, not all tools were a simple `go get` => `go install` change:

- `kustomize`: Incompatible with `go install` because its `go.mod` has
`exclude` and `replace` directives. The template now uses the official
install script.
- `controller-gen`: Fully compatible with `go install`.
- `setup-envtest`: Can be installed using `go install`, but not a
specific version since the repo does not have the tag scheme since it is
in a sub-module. It would need a tag in the format
`tools/setup-envtest/vX.Y.Z`.

The commit also regenerated all book source code using this new pattern.

Resolves: https://github.com/kubernetes-sigs/kubebuilder/issues/2477

### CODE REVIEW

File: pkg/cli/alpha/config-gen/examples/kustomize/README.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ generated are provided as `resources` input, the inputs will be modified by the
 Install the latest version of `kustomize`.
 
 ```sh
-GO111MODULE=on go get sigs.k8s.io/kustomize/kustomize/v4
```
Comment (camilamacedo86) [inline pkg/cli/alpha/config-gen/examples/kustomize/README.md:22]: cannot be go install?

File: pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go [hunk 2]
```diff
@@ -154,35 +154,31 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
...
 	$(KUSTOMIZE) build config/default | kubectl delete --ignore-not-found=$(ignore-not-found) -f -
 
-.PHONY: controller-gen
-controller-gen: ## Download controller-gen locally if necessary.
-ifeq (, $(shell which controller-gen))
-	@{ \
-	set -e ;\
-	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
-	cd $$CONTROLLER_GEN_TMP_DIR ;\
-	go mod init tmp ;\
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go:164]: could we not try to just replace go get by install? 
This shows that we are trying to do a refac in the file and it will make it harder for the users to upgrade the changes

File: testdata/project-v3-multigroup/config/crd/bases/ship.testproject.org_cruisers.yaml [hunk 3]
```diff
@@ -1,9 +1,10 @@
+
 ---
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.8.0
+    controller-gen.kubebuilder.io/version: v0.6.2
```
Comment (camilamacedo86) [inline testdata/project-v3-multigroup/config/crd/bases/ship.testproject.org_cruisers.yaml:7]: why we are downgrading this one?

File: pkg/cli/alpha/config-gen/examples/kustomize/README.md (hunk 1 above)
Comment (ryantking) [reply pkg/cli/alpha/config-gen/examples/kustomize/README.md:22]: No, explanation in PR description.

File: testdata/project-v3-multigroup/config/crd/bases/ship.testproject.org_cruisers.yaml (hunk 3 above)
Comment (ryantking) [reply testdata/project-v3-multigroup/config/crd/bases/ship.testproject.org_cruisers.yaml:7]: Temporary to make `make generate` work, details in PR description. Not sure what the fix for the problem I explained is, sure it's something simple that I'm missing.

File: pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go (hunk 2 above)
Comment (ryantking) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go:164]: No, explanation in PR description.

File: docs/book/src/cronjob-tutorial/testdata/project/Makefile [hunk 4]
```diff
@@ -81,29 +81,37 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
...
-CONTROLLER_GEN = $(shell pwd)/bin/controller-gen
-controller-gen: ## Download controller-gen locally if necessary.
-	$(call go-get-tool,$(CONTROLLER_GEN),sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.1)
-
-KUSTOMIZE = $(shell pwd)/bin/kustomize
-kustomize: ## Download kustomize locally if necessary.
-	$(call go-get-tool,$(KUSTOMIZE),sigs.k8s.io/kustomize/kustomize/v3@v3.8.7)
-
-ENVTEST = $(shell pwd)/bin/setup-envtest
-envtest: ## Download envtest-setup locally if necessary.
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/Makefile:94]: You will probably need to solve https://github.com/kubernetes-sigs/kubebuilder/issues/2480 in this PR.

File: pkg/cli/alpha/config-gen/examples/kustomize/README.md (hunk 1 above)
Comment (camilamacedo86) [reply pkg/cli/alpha/config-gen/examples/kustomize/README.md:22]: Hi @ryantking please see: https://github.com/kubernetes-sigs/kubebuilder/issues/2480

File: pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go (hunk 2 above)
Comment (camilamacedo86) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go:164]: Hi @ryantking please see: https://github.com/kubernetes-sigs/kubebuilder/issues/2480

File: pkg/plugins/golang/v3/scaffolds/init.go [hunk 5]
```diff
@@ -32,7 +32,7 @@ const (
 	// ControllerRuntimeVersion is the kubernetes-sigs/controller-runtime version to be used in the project
 	ControllerRuntimeVersion = "v0.11.0"
 	// ControllerToolsVersion is the kubernetes-sigs/controller-tools version to be used in the project
-	ControllerToolsVersion = "v0.8.0"
+	ControllerToolsVersion = "v0.6.2"
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/init.go:35]: could you please revert this one?

Comment (ryantking) [reply pkg/plugins/golang/v3/scaffolds/init.go:35]: Yes, `controller-gen` v0.8.0 breaks `testdata/projectv3-v1beta1`. Need to figure out how to set only that one to an older version or fix the compatibility issues. See the PR description for more details.

File: docs/book/src/cronjob-tutorial/testdata/project/config/crd/bases/batch.tutorial.kubebuilder.io_cronjobs.yaml [hunk 6]
```diff
@@ -4,7 +4,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.6.1
+    controller-gen.kubebuilder.io/version: v0.6.2
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/config/crd/bases/batch.tutorial.kubebuilder.io_cronjobs.yaml:7]: That shows wrong we must to update the docs samples and it should be using 0.8.0
We have an script to -generate all but we need to diff with master and add the missing parts that can be indeed a follow up

File: pkg/cli/alpha/config-gen/cmd.go [hunk 7]
```diff
@@ -254,7 +254,7 @@ kubebuilder alpha config-gen kubebuilderconfiggen.yaml patch1.yaml patch2.yaml
 #
 
 # install the latest kustomize
-GO111MODULE=on go get sigs.k8s.io/kustomize/kustomize/v4
+curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"  | bash
```
Comment (camilamacedo86) [inline pkg/cli/alpha/config-gen/cmd.go:257]: If we only do that we will install the latest version always. We need to pass the version to the script.

File: testdata/project-v3-config/tools.go [hunk 8]
```diff
@@ -0,0 +1,9 @@
+//go:build tools
+// +build tools
+
+// This package imports things required by build scripts, to force go mod to see them as dependencies
+package tools
+
+import (
+	_ "sigs.k8s.io/controller-runtime/tools/setup-envtest"
+)
```
Comment (camilamacedo86) [inline testdata/project-v3-config/tools.go:9]: cool 🥇

**Comment by mikebz** [comment]
ahh... saw this too late, but it looks like it's more evolved than my initial stab at this problem https://github.com/kubernetes-sigs/kubebuilder/pull/2496 - what do we need to do to get this over the line?  Happy to help on the kustomize end as well.

**Comment by ryantking** [comment]
@camilamacedo86 @mikebz sorry for the delay, I think the PR is in a good state to be accepted now.

File: pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go [hunk 9]
```diff
@@ -154,35 +154,29 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
...
-KUSTOMIZE=$(shell which kustomize)
-endif
+kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
+$(KUSTOMIZE):
+	curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN)
+
+.PHONY: controller-gen
+controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
+$(CONTROLLER_GEN):
+	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
```
Comment (varshaprasad96) [inline pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go:181]: if `controller-gen` is already installed in the required path, are we running `go install` again every time ?
Same with `kustomize`?

Comment (ryantking) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/makefile.go:181]: Yes the makefile target dependencies are setup such that it'll only be installed if not present.

File: pkg/cli/alpha/config-gen/cmd.go (hunk 7 above)
Comment (ryantking) [reply pkg/cli/alpha/config-gen/cmd.go:257]: The documentation line I'm replacing did not have a version, it just installed the latest v4 release. Should we add a specific version here?

**Comment by camilamacedo86** [comment]
Hi @ryantking, 

Note that the changes are not passing in the `netlify/kubebuilder/deploy-preview`. Also, we need to squash/rebase the commits could you please check these ones so we are able to move forward here?

File: docs/book/src/component-config-tutorial/testdata/project/api/v2/projectconfig_types.go [hunk 10]
```diff
@@ -18,25 +18,47 @@ package v2
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	cfg "sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/api/v2/projectconfig_types.go:21]: @ryantking , could you please add this line back?

File: docs/book/src/component-config-tutorial/testdata/project/api/v2/zz_generated.deepcopy.go [hunk 11]
```diff
@@ -1,51 +0,0 @@
-// +build !ignore_autogenerated
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/api/v2/zz_generated.deepcopy.go:1]: After update the projects in the docs, could you please call the target make all to ensure that we will update/generate all manifest?

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go [hunk 12]
```diff
@@ -13,160 +13,36 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
 import (
-	batchv1beta1 "k8s.io/api/batch/v1beta1"
-	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
-// +kubebuilder:docs-gen:collapse=Imports
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go:33]: @ryantking , after run the scripts we need to compare with master and ensure that we will add back all code/comments and etc on top. Could you please add back all docs that were removed?

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go [hunk 13]
```diff
@@ -13,160 +13,36 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
 
-/*
- */
 package v1
 
-/*
- */
 import (
-	batchv1beta1 "k8s.io/api/batch/v1beta1"
-	corev1 "k8s.io/api/core/v1"
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go:26]: Same here. It needs be back

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go [hunk 14]
```diff
@@ -136,78 +73,3 @@ func (r *CronJob) ValidateDelete() error {
...
-		// a 11-character suffix to the cronjob (`-$TIMESTAMP`) when creating
-		// a job. The job name length limit is 63 characters. Therefore cronjob
-		// names must have length <= 63-11=52. If we don't validate this here,
-		// then job creation will fail later.
-		return field.Invalid(field.NewPath("metadata").Child("name"), r.Name, "must be no more than 52 characters")
-	}
-	return nil
-}
-
-// +kubebuilder:docs-gen:collapse=Validate object name
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go:213]: same here.

File: docs/book/src/component-config-tutorial/testdata/project/main.go [hunk 15]
```diff
@@ -62,10 +62,9 @@ func main() {
 	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
 
 	var err error
-	ctrlConfig := configv2.ProjectConfig{}
 	options := ctrl.Options{Scheme: scheme}
 	if configFile != "" {
-		options, err = options.AndFrom(ctrl.ConfigFile().AtPath(configFile).OfKind(&ctrlConfig))
+		options, err = options.AndFrom(ctrl.ConfigFile().AtPath(configFile))
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/main.go:67]: The source/code here, the example from the tutorial was changed.
Could you please revert?

File: pkg/cli/alpha/config-gen/cmd.go (hunk 7 above)
Comment (camilamacedo86) [reply pkg/cli/alpha/config-gen/cmd.go:257]: Yes. We need to clear to the users that they ought to use the compatible/supported version. 
E.g: https://github.com/kubernetes-sigs/kubebuilder/pull/2486/files#diff-2ce1a29a9c5ec9b4069ab522cad03d59a54521b8311fcaab953a551be8aa5b23R116

See: https://book.kubebuilder.io/quick-start.html#versions-and-supportability
Screenshot 2022-02-04 at 17 39 28

In this case e.g. if users use v4 for kustomize then, it schema will not work with k8s versions < 1.22. See: https://github.com/kubernetes-sigs/kubebuilder/pull/2498#issuecomment-1026772984

**Review by camilamacedo86 — COMMENTED** [review]
Very close just a few nits now.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go [hunk 16]
```diff
@@ -159,32 +159,35 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
...
+KUSTOMIZE_VERSION ?= {{ .KustomizeVersion }}
+CONTROLLER_TOOLS_VERSION ?= {{ .ControllerToolsVersion }}
 
-KUSTOMIZE = $(shell pwd)/bin/kustomize
+KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
 .PHONY: kustomize
-kustomize: ## Download kustomize locally if necessary.
-	$(call go-get-tool,$(KUSTOMIZE),sigs.k8s.io/kustomize/kustomize/v3@{{ .KustomizeVersion }})
+kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
+$(KUSTOMIZE):
```
Comment (bprashanth) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:181]: Does this need to be `$(KUSTOMIZE): $(LOCALBIN)` ?

File: docs/book/src/component-config-tutorial/testdata/project/Makefile [hunk 17]
```diff
@@ -32,78 +33,101 @@ all: build
...
+
+.PHONY: controller-gen
+controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
+$(CONTROLLER_GEN):
+	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+
+.PHONY: envtest
+envtest: $(ENVTEST) ## Download envtest-setup locally if necessary.
+$(ENVTEST):
+	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-runtime/tools/setup-envtest
```
[the remark this answers is not in the document]
Comment (ryantking) [reply docs/book/src/component-config-tutorial/testdata/project/Makefile:133]: @bprashanth The command will install the version locked in the `go.sum` from the `tools.go` file. More details in the PR description

**Review by camilamacedo86 — APPROVED** [review]
Great work 🥇 
It shows fine for me and have my 
/approved

Also, I see that you updated the docs and their samples. 🥇 Terrific
it would be nice get a second look/reviewer as well.

File: docs/book/src/cronjob-tutorial/testdata/project/Makefile [hunk 18]
```diff
@@ -32,78 +33,101 @@ all: build
...
+## Tool Binaries
+KUSTOMIZE ?= $(LOCALBIN)/kustomize
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
+ENVTEST ?= $(LOCALBIN)/setup-envtest
+
+## Tool Versions
+KUSTOMIZE_VERSION ?= v3.8.7
+CONTROLLER_TOOLS_VERSION ?= v0.8.0
+
+KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
```
Comment (jberkhahn) [inline docs/book/src/cronjob-tutorial/testdata/project/Makefile:119]: not sure how kustomize does releases,  but this looks like you're grabbing it off master, do they not have release tags or something so you could point to a consistent version?

**Comment by jberkhahn** [comment]
I actually deleted that comment because I went and looked at how the script works, you provide a version # and it installs that version (which is what we do here, we always install v3.8.7 unless it's been overidden by a pre-existing env var). I think it's fine?

File: docs/book/install-and-build.sh [hunk 19]
```diff
@@ -67,11 +67,11 @@ ${cmd} /tmp/mdbook.${ext}
 chmod +x /tmp/mdbook
 
 echo "grabbing the latest released controller-gen"
-go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.5.0
+go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.5.0
```
Comment (wmuizelaar) [inline docs/book/install-and-build.sh:70]: shouldn't this be v0.8.0 to be consistent with the rest of the used versions?

File: docs/book/src/component-config-tutorial/testdata/project/tools.go [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+limitations under the License.
+*/
+// +kubebuilder:docs-gen:collapse=Apache License
+
+// This package imports things required by build scripts, to force go mod to see them as dependencies
+package tools
+
+import (
+	_ "sigs.k8s.io/controller-runtime/tools/setup-envtest"
+)
```
Comment (joelanford) [inline docs/book/src/component-config-tutorial/testdata/project/tools.go:25]: This setup is going to result in the project's CI dependencies being added to the main project's go.mod file, which has implications on:
- how underlying dependencies get resolved when a build tool and the main project share a transitive dependency (e.g. the build tool might unnecessarily hold back the transitive dependency on a lower version due to Go's MVS algorithm)
- how other projects consume the main project's as a dependency (external projects will end up with this project's build tool dependencies in their go.mod files and might impact their transitive dependency selection also due to MVS)

TL;DR: At a minimum, I think we need to put the tools dependencies in a separate go module. However see also this issue for background about why even that might not be a great idea: https://github.com/kubernetes-sigs/cluster-api/issues/5569

/hold

**Comment by joelanford** [comment]
> we are only trying to use this approach with the env-test

That may be, but many users will see this tools pattern and start adding other build dependencies, thus further polluting their main `go.mod` file. One of the main goals of kubebuilder is to help reinforce best practices, and I'd consider it a bad practice to include build deps alongside library deps in the same go.mod/go.sum.

**Comment by joelanford** [comment]
One way to solve this problem is something like:
```sh
CR_VERSION=$(shell go list -m sigs.k8s.io/controller-runtime | cut -d" " -f2)
ENVTEST = $(shell pwd)/bin/setup-envtest
.PHONY: envtest
envtest: ## Download envtest-setup locally if necessary.
	$(call go-get-tool,$(ENVTEST),sigs.k8s.io/controller-runtime/tools/setup-envtest@$(CR_VERSION))

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

That does not seems to work, see: 

```
 go get sigs.k8s.io/controller-runtime/tools/setup-envtest@v0.11.1
go: downloading sigs.k8s.io/controller-runtime v0.11.1
go get: module sigs.k8s.io/controller-runtime@v0.11.1 found, but does not contain package sigs.k8s.io/controller-runtime/tools/setup-envtest
```

 Also, see: https://github.com/kubernetes-sigs/controller-runtime/issues/1670

@ryantking, change. the env-test to use the latest shows to be beyond of the scope of this PR, so, wdyt about we keep it still using latest to get this PR merged and discuss it further in  https://github.com/kubernetes-sigs/kubebuilder/issues/2480?

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

Do you see any other reason for this one did not getting merged now? WDYT? 
Can we move forward with?

**Review by jmrodri — APPROVED** [review]
/lgtm

**Review by rashmigottipati — APPROVED** [review]
